---
title: "RUNX2 KO"
author: "Alex"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
html_document:
toc: TRUE
toc_float: FALSE
editor_options: 
  chunk_output_type: console
chunk_output_type: console
---


<style type="text/css">
.main-container {
max-width: 2500px;
margin-left: 5px;
margin-right: auto;
}
.toc-content {
max-width: 2500px;
margin-left: 50px;
margin-right: auto;
}

div {

margin-left: 5px;
}


hr.new1 {
border-top: 1px solid #84a8e0;
}


</style>

version: 1.0 <br />
Run at "`r format(Sys.time())`"

# Data {.tabset}

## Overview/Method:

  * RUNX2 was identified as a differential TF for ATAC-seq cluster 1. See ATAC-seq exp 2a. 
  * We are interested on the downstream regulation cascade of RUNX2. 
  * The knock-outs (KO) were generated using a single RNA guide (sgRNA) from Synthego.
    + The sgRNA targeting RUNX2 is CCGUAGAUCCGAGCACCAGC.						
    + And the sgNEG is a non-targeting guide.						
    + Both cell lines are expressing Cas9 constitutively. 						
    + The experiment was done using the sgRNA and RNAiMAX lipofectamine reagent. 						
  						
  * While running QC a strong batch was detected between cell lines.  Thus cpm was batch corrected with limma removeBatchEffect and cell line was added as a covariate in the limma-model
  
  
  
```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, cache=F}

knitr::opts_chunk$set(include=F, echo=FALSE, message=FALSE, warning=FALSE, fig.show="asis", fig.keep="all",tidy.opts=list(width.cutoff=100),tidy=TRUE)
options(knitr.table.format = "latex")
options(width = 1600)


### global options 
options(scipen=999)
### standard libraries
library("RColorBrewer")
library(dplyr)
library(openxlsx)
library ( parallelDist)
library(forcats)
library ( factoextra)
library ( FactoMineR)
library ( ggplot2)
library ( ggplotify) 
library ( patchwork)
library ( limma)
library ( edgeR)
library(knitr)        
library ( kableExtra)    
library(tidyr)
library ( data.table )
library (DT)
library ( ComplexHeatmap)
getPalette = colorRampPalette(brewer.pal(9, "Set1")) # expand color pallete
# keeping things consistent
ss = 123
set.seed(ss)



## read in resources 
## 
main.data = "/data/"
resource = "/ehome/resource/"
resource.ext = "/ehome/"


source ( "/ehome/scripts/config.ext/ext_June.2020.R" )

# for info on genes
genecard  <- readRDS(  paste0( resource, "annotation/genecard.rds" )  )
genecard = genecard[ , c("InputTerm","EntrezGene")]
genecard = genecard[!is.na(genecard$EntrezGene), ]
colnames(genecard) = c("GENE","description")

## pathways 
pathways = readRDS( paste0( resource, "/gsea/limma/go_path.rds"))
# change column to make it work with profiler 
pathway.p = pathways
colnames ( pathway.p)[1] = c("geneID")

annt =  readRDS( paste0( resource, "/gsea/limma/go_path_withWeneName.rds"))
annt$GeneID = NULL 

colnames ( annt ) = c ( "gene"  , "term")
annt = annt [ ,  c ( "term"  , "gene") ]
pathgsea =  readRDS( paste0( resource, "/gsea/limma/go_path_fgsea_genename.rds"))


# loading up cancer genes
cosmic = read.csv("https://www.dropbox.com/s/ceo0ol2h9vt01c3/cosmic-cancer_gene_census.v85.csv?dl=1")
tsg = cosmic[ grepl("TSG", cosmic$Role.in.Cancer, ignore.case = T), ]
oncogene = cosmic[ grepl("oncog", cosmic$Role.in.Cancer, ignore.case = T), ]
cosmic$Gene.Symbol = as.character(cosmic$Gene.Symbol)

# load drug database 
drug.db = "https://www.dropbox.com/s/1k60wzvstbp0m7w/7.8.2020.xlsx?dl=1"
drug.db <- read.xlsx(drug.db , colNames = TRUE)
drug.db = drug.db[ grepl ( "antagonist|blocker|antibody|antisense|inhibit", drug.db$interaction_types), ]



# main excel output for later 
wb <- createWorkbook()

out.dir = "./results.out/"
dir.create(out.dir)


```


```{r}
# get key. 

key1 =  read.xlsx( "https://www.dropbox.com/s/n13y22wslxp7kwo/EF_RNAseq_Samples_StanEdit.xlsx?dl=1")
key = key1 [ order ( key1$sgRNA), ]

# get starcount 
data =  read.table("../SUMMARY/counts/raw.tsv")
# find samples that were removed 
rm = setdiff ( key$RNAseq.id, colnames ( data ))
rm = which ( key$RNAseq.id %in% rm )

```



## QC & Key {.tabset}


### Table Key

```{r, message=FALSE, warning=FALSE,fig=TRUE,fig.width=16, fig.height=10, echo=FALSE, include=TRUE}


kable( key  , format = "html" , row.names = F, caption = "Main Key" ) %>% kable_classic(full_width = F, position = "center")  %>% 
  row_spec(row= rm
           , bold = F 
           , color = "black"
           # , background = "#c6c9cf"
           ,strikeout = TRUE
           ) %>%
  footnote(
           symbol  = c("strikeout indicates that sample was not sequenced. Please see Seq notes for more details"),
         
           )

key = key [ -rm, ]


```

```{r}


key$tube = key$RNAseq.id 
data = data[ , key$RNAseq.id]
all.equal ( colnames ( data), key$RNAseq.id ) # sanity check
 
 
 
## inititiate edgeR 
y.match <- DGEList( data)

## remove low counts since this could interfere with some of the statistical approximations
#  here half the total sample size must have > 2 raw to be included 
# adjust as needed 

threshold <- ceiling(  ncol ( data )/2  )
mincpm = 1
keep <- rowSums(  cpm ( data, log=F ) >  mincpm ) >= threshold


# do we need to force a gene? 
fgene = "RUNX2"
keep [fgene]
keep [fgene] = TRUE 

# before 
dim(y.match)
y.match  <- y.match [keep, ]
# after 
dim (y.match)    

## normalisation using trimmed mean of M-values (TMM) (Robinson and Oshlack 2010)
## in additional to library scaling this will ensure that the expression distributions of each sample are similar
## across the experiment
y.match <- calcNormFactors( y.match, method = "TMM" )

# convert this for further downstream but log + 1 

cpm = log2  ( as.data.frame(cpm(y.match, prior.count=3, log=F)) + 1 )

all.equal(names(cpm), key$tube ) 


```


```{r}
# set up model for limma-voom 
# retrospectively this experiment needs some batch correction 

group = key$sgRNA 
cell = key$Cell.line
design <- model.matrix(~0  +  group + cell  )
colnames ( design) = gsub ( 'group',"",colnames(design))

voom <- voom(y.match, plot = T, design = design)  
voom.plot <- recordPlot()

```

```{r}
# this experiment needs batch correction 

# after batch correction 
#modcombat <- model.matrix (~1, data=key)
#cpm_mod = ComBat(dat=as.matrix(cpm ), batch=key$Cell.line, mod=modcombat, par.prior=TRUE, prior.plots=FALSE)

#cpm_mod = cpm_mod [ , key$RNAseq.id]
#cpm_mod = data.frame ( cpm_mod )

cpm_mod2 <- removeBatchEffect(cpm, key$Cell.line)
cpm_mod2 <- data.frame ( cpm_mod2 )

# limma's batch correction is better 

cpm_mod = cpm_mod2 [ , key$RNAseq.id]


```

### Density plot of log2 normalize counts

* To assess normalization we plot the densities of counts for the each samples. A succesful normalization will lead to overlapping densities.

```{r, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=6}

density1 = melt ( cpm_mod  )
colnames(density1) = c("sample", "value")
density1 = ggplot(density1, aes(value, fill=sample, colours=sample)) + geom_density(alpha=0.1) + 
      theme_bw()  + scale_color_manual( values= getPalette(12)) + ggtitle ( "after correction")

density1



bx.result = box.outlier ( cpm, key, group.id= "sgRNA", tname="tube" ) 
bx.result2 = box.outlier ( cpm, key, group.id= "Cell.line", tname="tube" ) 

```

### Correlation Matrix 

* We can cleary see here that the thre is a strong batch effect with the cell line. 
* thus a batch correction must be implemented. 
* cor-cor plot: x-axis by sample. y-axis is the coeffecient correlation between the sample vs other.  
  + color coded by group
  + the expected correlation between samples, irrespective of group should be ~ r= 0.9


```{r}

bx.result3 = box.outlier ( cpm_mod, key, group.id= "Cell.line", tname="tube" ) 


bx.result2 = bx.result2$g1 + ggtitle ( "Before correction")
bx.result3 = bx.result3$g1 + ggtitle ( "After correction")

# bring the results range back in 

bx.result3 = bx.result3 + ylim(.8,1) 


```


```{r, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=6}
bx.result2 + bx.result3
```

### mean variance 

* Mean variance trend plot for VOOM stabilization
* https://f1000research.com/articles/5-1408
* What we want here is to show a decreasing trend between the means and variances 


```{r, message=FALSE, warning=FALSE,fig=TRUE,fig.width=8, fig.height=5, echo=FALSE, include=TRUE}
voom.plot
```


### Single genes {.tabset}

  * This is important to ascertain expected gene expectation.
  * pv is determine by wilcox test

```{r, message=FALSE, warning=FALSE,fig=TRUE,fig.width=5, fig.height=8, echo=FALSE, include=TRUE, results='asis'}

key$group = key$sgRNA

for ( gk in unique ( c ( "GAPDH", "ACTB", "RUNX2" ) ) ){
  
 temp <- melt (as.matrix ( cpm_mod [row.names( cpm_mod ) %in% gk,]  )  )
    colnames ( temp ) = c("gene","tube","value")
 temp = merge ( temp, key, by="tube") 
    
colsingl <- getPalette ( length (  unique ( temp$sgRNA  )  )) 
temp$group = factor ( temp$group, levels=unique (sort ( temp$group) ) )
    
g1 = ggplot(temp, aes(x= group , y=value )) +
        geom_violin()+ 
        geom_jitter(shape=19, position=position_jitter(0.07), aes( colour=sgRNA) , size = 5, alpha= .5) +
        theme_bw() +
        ylab( "log2 ( cpm + 1 )") +
        xlab("")  +
        theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
              #axis.text.x = element_blank(),
              #axis.text.y = element_blank(), # comment this out to display cancer.subtype
              axis.text.y = element_text(size=12),
              axis.text.x = element_text(size=15, angle = 40,  hjust=1 ),
              axis.title.x = element_text(size=22),
              axis.title.y     = element_text(size=22), 
              legend.text      =element_text(size=12)
        ) + scale_color_manual(values = unique (colsingl )) +
        stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
                     geom = "crossbar", width = .5) + ggtitle ( gk )


g1 = ggpval::add_pval(g1, pairs = list(c(1, 2)), test='wilcox.test')

dis = paste0 ( "#### ", gk, "<br> \n")
cat(dis ) 
print ( g1 )

cat('\n\n')

  
  
  # end gene 
  # 
  cat ( "\n")



}


```

```{r}

# run cluster analysis to see how well the experiment looks from a global perspective. 

this.color = c( RUNX2="#459fcc", sgNeg="#b58953") 

distuse = "euclidean"
linkuse = "ward.D2"

key_c = key
m.cpm = cpm_mod2  [ , key_c$tube] 


all.equal(names(m.cpm), key_c$tube ) 

sd <- apply (m.cpm, 1,  function(x) cov(x)  )

p = c(.1,.2,.3,.4,.5,.6, .7, .75, .8, .9,.95)
q <- quantile (as.numeric ( sd ), probs = p , na.rm =T )

# correct for variance
get.p = "95%"
get.p = as.numeric ( q[get.p] )
sd2 = sd[ !is.na(sd) & sd >get.p  ]

# from here on out cm.cpm2 is the main one since cm.cpm changes 
m.cpm = m.cpm [ row.names ( m.cpm) %in% names( sd2), ]
m.cpm = setcolorder(m.cpm,  as.character( key_c$tube )  )




dend2 = makehr ( df.this = t( m.cpm ), km=2, dist.this = distuse, aggreg=linkuse, meta1 = key_c$group  )    

dend2$dend = dend2$dend %>% set("labels_cex", 1) %>% set("labels_col", "black")

# make colors 
set.seed(123)
colqc <- getPalette ( 50 ) 

cond.color =   c("#fcc203", "#4dd413" )
names ( cond.color ) = unique ( key_c$group) 





par(mar=c(5,5,2,10))
# c(bottom, left, top, right)
plot ( dend2$dend, horiz=T) 
colored_bars(colors =cond.color[ key_c$group], dend =  dend2$dend, sort_by_labels_order = TRUE, horiz = T, text_shift=0, x_scale=0)

legend("topleft", legend=c ( names ( cond.color ) ) 
       ,lwd=8,  box.col = "white", bg = "white", border.col = "white",
       col=c ( as.character(cond.color) ) , lty=1, cex=1)
                  
hc.all = recordPlot()
 
 pca.biop <- PCA( t( m.cpm ), graph = FALSE)

p <- fviz_pca_ind(pca.biop, label="none", addEllipses=TRUE, ellipse.level=0.95,  axes = c(1, 2),  alpha=0  )
p2 <- p +geom_point( aes(colour=  key_c$sgRNA , shape=factor ( key_c$Cell.line ) ), size=9.0 , alpha=.4 )



pca.biop = p2  + scale_colour_manual(values = this.color, name  ="group" ) +
 scale_shape_discrete ( name="Cell Line") + ggtitle("Unsupervised") 



```


### Unsupervised PCA

* top 95 percentile most variant genes ( determine with coefficient of variation  )

```{r, message=FALSE, warning=FALSE,fig=TRUE,fig.width=10, fig.height=7, echo=FALSE, include=TRUE}
pca.biop
```


```{r}
gsva = 0 

if ( gsva == 1 ){
# run GSVA 
gsv_pathways = readRDS( paste0( resource, "/gsea/limma/go_path_fgsea_genename.rds"))
library ( GSVA )
gsva.plate <- GSVA::gsva(as.matrix ( cpm_mod ), gsv_pathways, min.sz=30, max.sz=500000
                   , mx.diff=TRUE, verbose=FALSE
                   , parallel.sz=1
                   ,  kcdf="Gaussian"
                   , method = "gsva")
saveRDS( gsva.plate , "GSVA_wholeset")



}else {
  gsva.plate = readRDS( "GSVA_wholeset" )

}


#gsva.platem = removeBatchEffect(gsva.plate, batch=key$Cell_line ) 

 

```














