---
title: "RUNX1 KO"
author: "Alex"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
html_document:
toc: TRUE
toc_float: FALSE
editor_options: 
  chunk_output_type: console
chunk_output_type: console
---


<style type="text/css">
.main-container {
max-width: 2500px;
margin-left: 5px;
margin-right: auto;
}
.toc-content {
max-width: 2500px;
margin-left: 50px;
margin-right: auto;
}

div {

margin-left: 5px;
}


hr.new1 {
border-top: 1px solid #84a8e0;
}


</style>

version: 1.0 <br />
Run at "`r format(Sys.time())`"

# Data {.tabset}

## Overview/Method:

  * RUNX2 was identified as a differential TF for ATAC-seq cluster 1. See ATAC-seq exp 2a. 
  * We are interested on the downstream regulation cascade of RUNX2. 
  * The knock-outs (KO) were generated using a single RNA guide (sgRNA) from Synthego.
    + The sgRNA targeting RUNX2 is CCGUAGAUCCGAGCACCAGC.						
    + And the sgNEG is a non-targeting guide.						
    + Both cell lines are expressing Cas9 constitutively. 						
    + The experiment was done using the sgRNA and RNAiMAX lipofectamine reagent. 						
  						

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, cache=F}

knitr::opts_chunk$set(include=F, echo=FALSE, message=FALSE, warning=FALSE, fig.show="asis", fig.keep="all",tidy.opts=list(width.cutoff=100),tidy=TRUE)
options(knitr.table.format = "latex")
options(width = 1600)


### global options 
options(scipen=999)
### standard libraries
library("RColorBrewer")
library(dplyr)
library(openxlsx)
library ( parallelDist)
library(forcats)
library ( factoextra)
library ( FactoMineR)
library ( ggplot2)
library ( ggplotify) 
library ( limma)
library ( edgeR)
library(knitr)        
library ( kableExtra)    
library(tidyr)
library ( data.table )
library (DT)
library ( ComplexHeatmap)
getPalette = colorRampPalette(brewer.pal(9, "Set1")) # expand color pallete
# keeping things consistent
ss = 123
set.seed(ss)



## read in resources 
## 
main.data = "/data/"
resource = "/ehome/resource/"
resource.ext = "/ehome/"


source ( "/ehome/scripts/config.ext/ext_June.2020.R" )

# for info on genes
genecard  <- readRDS(  paste0( resource, "annotation/genecard.rds" )  )
genecard = genecard[ , c("InputTerm","EntrezGene")]
genecard = genecard[!is.na(genecard$EntrezGene), ]
colnames(genecard) = c("GENE","description")

## pathways 
pathways = readRDS( paste0( resource, "/gsea/limma/go_path.rds"))
# change column to make it work with profiler 
pathway.p = pathways
colnames ( pathway.p)[1] = c("geneID")

annt =  readRDS( paste0( resource, "/gsea/limma/go_path_withWeneName.rds"))
annt$GeneID = NULL 

colnames ( annt ) = c ( "gene"  , "term")
annt = annt [ ,  c ( "term"  , "gene") ]
pathgsea =  readRDS( paste0( resource, "/gsea/limma/go_path_fgsea_genename.rds"))


# loading up cancer genes
cosmic = read.csv("https://www.dropbox.com/s/ceo0ol2h9vt01c3/cosmic-cancer_gene_census.v85.csv?dl=1")
tsg = cosmic[ grepl("TSG", cosmic$Role.in.Cancer, ignore.case = T), ]
oncogene = cosmic[ grepl("oncog", cosmic$Role.in.Cancer, ignore.case = T), ]
cosmic$Gene.Symbol = as.character(cosmic$Gene.Symbol)

# load drug database 
drug.db = "https://www.dropbox.com/s/1k60wzvstbp0m7w/7.8.2020.xlsx?dl=1"
drug.db <- read.xlsx(drug.db , colNames = TRUE)
drug.db = drug.db[ grepl ( "antagonist|blocker|antibody|antisense|inhibit", drug.db$interaction_types), ]



# main excel output for later 
wb <- createWorkbook()

out.dir = "./results.out/"
dir.create(out.dir)


```


```{r}
# get key. 

key1 =  read.xlsx( "https://www.dropbox.com/s/n13y22wslxp7kwo/EF_RNAseq_Samples_StanEdit.xlsx?dl=1")
key = key1 [ order ( key1$sgRNA), ]

# get starcount 
data =  read.table("../SUMMARY/counts/raw.tsv")
# find samples that were removed 
rm = setdiff ( key$RNAseq.id, colnames ( data ))
rm = which ( key$RNAseq.id %in% rm )

```



## QC & Key {.tabset}


### Table Key

```{r, message=FALSE, warning=FALSE,fig=TRUE,fig.width=16, fig.height=10, echo=FALSE, include=TRUE}


kable( key  , format = "html" , row.names = F, caption = "Main Key" ) %>% kable_classic(full_width = F, position = "center")  %>% 
  row_spec(row= rm
           , bold = F 
           , color = "black"
           # , background = "#c6c9cf"
           ,strikeout = TRUE
           ) %>%
  footnote(
           symbol  = c("strikeout indicates that sample was not sequenced. Please see Seq notes for more details"),
         
           )

key = key [ -rm, ]


```

```{r}


key$tube = key$RNAseq.id 
data = data[ , key$RNAseq.id]
all.equal ( colnames ( data), key$RNAseq.id ) # sanity check
 
 
 
## inititiate edgeR 
y.match <- DGEList( data)

## remove low counts since this could interfere with some of the statistical approximations
#  here half the total sample size must have > 2 raw to be included 
# adjust as needed 

threshold <- ceiling(  ncol ( data )/2  )
mincpm = 1
keep <- rowSums(  cpm ( data, log=F ) >  mincpm ) >= threshold


# do we need to force a gene? 
fgene = "RUNX2"
keep [fgene]
keep [fgene] = TRUE 

# before 
dim(y.match)
y.match  <- y.match [keep, ]
# after 
dim (y.match)    

## normalisation using trimmed mean of M-values (TMM) (Robinson and Oshlack 2010)
## in additional to library scaling this will ensure that the expression distributions of each sample are similar
## across the experiment
y.match <- calcNormFactors( y.match, method = "TMM" )

# convert this for further downstream but log + 1 
cpm = as.data.frame(cpm(y.match, prior.count=3, log=T))
cpm.log2p = log2  ( as.data.frame(cpm(y.match, prior.count=3, log=F)) + 1 )

all.equal(names(cpm), key$tube ) 


```


```{r}
# set up model for limma-voom 

group = key$sgRNA 

design <- model.matrix(~0  +  group )
colnames ( design) = gsub ( 'group',"",colnames(design))

voom <- voom(y.match, plot = T, design = design)  
voom.plot <- recordPlot()

```






